### PR TITLE
Add ffmpeg package

### PR DESCRIPTION
**Pull request by uberhacker**
FFmpeg is the leading multimedia framework, able to decode, encode, transcode, mux, demux, stream, filter and play pretty much anything that humans and machines have created.  See http://ffmpeg.org/. 
 ~This package is using the static build available at https://johnvansickle.com/ffmpeg/ which avoids the lengthy compile process to install it .~

### CODE REVIEW

**Comment by jam7** [comment]
Not working on armv7l.
```
$ uname -m
armv7l
$ crew install ffmpeg
...
Building from source, this may take a while...
Preconfiguring package...
Installing...
Ffmpeg installed!
$ ffmpeg
bash: /usr/local/bin/ffmpeg: cannot execute binary file: Exec format error
$ ffprobe
bash: /usr/local/bin/ffprobe: cannot execute binary file: Exec format error
```

**Comment by uberhacker** [comment]
@jam7:  Right, this version only works on x86_64 architecture.  You are welcome to compile from source and submit a new PR.  Warning: Not for the faint of heart. <3

**Comment by jam7** [comment]
If you want to make a package file for binary distribution, please use https://github.com/skycocker/chromebrew/pull/523#issuecomment-280819556 trick.  You can leave source_url and sha1 lines for only x86_64.

Without it, crew install x86_64 binary for armv7l or i686 although it doesn't work.

**Comment by lyxell** [comment]
@uberhacker Nice! Which platforms is this tested on?

**Comment by uberhacker** [comment]
@lyxell: Currently, only the x86_64 binary works.  I'm working on a complete solution that compiles from source but it has been a huge challenge so far.

**Comment by uberhacker** [comment]
@lyxell: Now that you merged in #682, it compiles cleanly on my x86_64 Chromebook, although it takes an eternity with all the dependencies. :) @jam7: Let me know if you can get it compiled on your armv7l system.

**Comment by jam7** [comment]
thanks i'll try it out.

File: packages/libass.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+class Libass < Package
+  version '0.13.6'
+  source_url 'https://github.com/libass/libass/archive/0.13.6.tar.gz'
+  source_sha1 '10fc35f8aba2a4a37aaed482a4184d144defb3d8'
+
+  depends_on 'freetype'
+  depends_on 'fribidi'
+  depends_on 'fontconfig'
+
```
Comment (jam7) [inline packages/libass.rb]: please add
```
depends_on 'automake'
depends_on 'autoconf'
```

File: packages/libfdkaac.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
+require 'package'
+
+class Libfdkaac < Package
+  version '0.1.5'
+  source_url 'https://github.com/mstorsjo/fdk-aac/archive/v0.1.5.tar.gz'
+  source_sha1 '2894bd6d21066bfdf65a2f8cb6d2ca2dde597194'
+
```
Comment (jam7) [inline packages/libfdkaac.rb]: please add
```
depends_on 'automake'
```

**Comment by jam7** [comment]
Unfortunately, libvpx is not installable on armv7l.
```
Building from source, this may take a while...
  enabling shared
  enabling vp8
  enabling vp9
  enabling postproc
  enabling vp9_postproc
  enabling webm_io
Unrecognized toolchain 'armv7l-linux-gcc'

Configuration failed. This could reflect a misconfiguration of your
toolchains, improper options selected, or another problem. If you
don't see any useful error messages above, the next step is to look
at the configure error log file (config.log) to determine what
configure was trying to do when it died.
libvpx failed to build: `./configure --target=armv7l-linux-gcc             --enable-shared             --enable-vp8             --enable-vp9             --enable-postproc             --enable-vp9-postproc             --enable-webm-io` exited with 1
```
So I don't test ffmpeg.  But, if this and ffmpeg work on x86_64, we have no problem to merge this I think.

**Comment by uberhacker** [comment]
@jam7: Thank you for taking the time to attempt a build of this on your system.  What happens if you just target the specific architecture (ie. `--target=#{ARCH}`)?

**Comment by uberhacker** [comment]
@jam7: Actually, looking at the list of supported targets, it appears that `--target=armv7-linux-gcc` might be you best option.  You might also want to try removing the `--target` configure option altogether and see what happens.
```
Supported targets:
    arm64-android-gcc        arm64-darwin-gcc         arm64-linux-gcc         
    armv7-android-gcc        armv7-darwin-gcc         armv7-linux-rvct        
    armv7-linux-gcc          armv7-none-rvct          armv7-win32-vs11        
    armv7-win32-vs12         armv7-win32-vs14        
    armv7s-darwin-gcc       
    armv8-linux-gcc         
    mips32-linux-gcc        
    mips64-linux-gcc        
    sparc-solaris-gcc       
    x86-android-gcc          x86-darwin8-gcc          x86-darwin8-icc         
    x86-darwin9-gcc          x86-darwin9-icc          x86-darwin10-gcc        
    x86-darwin11-gcc         x86-darwin12-gcc         x86-darwin13-gcc        
    x86-darwin14-gcc         x86-darwin15-gcc         x86-iphonesimulator-gcc 
    x86-linux-gcc            x86-linux-icc            x86-os2-gcc             
    x86-solaris-gcc          x86-win32-gcc            x86-win32-vs10          
    x86-win32-vs11           x86-win32-vs12           x86-win32-vs14          
    x86_64-android-gcc       x86_64-darwin9-gcc       x86_64-darwin10-gcc     
    x86_64-darwin11-gcc      x86_64-darwin12-gcc      x86_64-darwin13-gcc     
    x86_64-darwin14-gcc      x86_64-darwin15-gcc      x86_64-iphonesimulator-gcc
    x86_64-linux-gcc         x86_64-linux-icc         x86_64-solaris-gcc      
    x86_64-win64-gcc         x86_64-win64-vs10        x86_64-win64-vs11       
    x86_64-win64-vs12        x86_64-win64-vs14       
    generic-gnu             
```
Thanks

**Comment by uberhacker** [comment]
@jam7: BTW, I also committed the missing package dependencies as you suggested.

**Comment by jam7** [comment]
It looks like not specifying `--target` seems good idea since that option is for cross-compile, AFAIK.  Without `--target` option, `configure` recognize target architecture much better.  However, still having the similar error, though.  It is saying:
```
gcc -march=armv7-a -mfloat-abi=softfp -mfpu=neon -DNDEBUG -O3 -fPIC -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=0 -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -Wall -Wdeclaration-after-statement -Wdisabled-optimization -Wfloat-conversion -Wpointer-arith -Wtype-limits -Wcast-qual -Wvla -Wimplicit-function-declaration -Wuninitialized -Wunused -Wextra -Wundef -E -o /tmp/vpx-conf-5221-7444.o /tmp/vpx-conf-5221-7444.c
In file included from /usr/local/include/features.h:392:0,
                 from /usr/local/include/stdio.h:27,
                 from /tmp/vpx-conf-5221-7444.c:1:
/usr/local/include/gnu/stubs.h:7:29: fatal error: gnu/stubs-soft.h: No such file or directory
 # include <gnu/stubs-soft.h>
                             ^
compilation terminated.
```
The source of this error is hard/soft mismatch.  Toolchain for armv7l provides only hardfp ABI.  But above command line requires soft ABI which is not present.

**Comment by uberhacker** [comment]
@jam7:  Just out of curiosity, what is the output of the detected target if you omit the `--target` option in the `configure` line?  You should see it only a few lines into the compile process.  Below is what I get:
```
Configuring for target 'x86_64-linux-gcc'
```

**Comment by jam7** [comment]
It says `armv7l-linux-gcc`.

Above error looks like common problem for armhf.  Here is some discussion.  https://bugzilla.mozilla.org/show_bug.cgi?id=1194905

**Comment by uberhacker** [comment]
Fixed with this patch?  https://hg.mozilla.org/mozilla-central/rev/2db4b0ef03a7

**Comment by uberhacker** [comment]
@jam7:  What happens if you compile without libvpx?  That library is more of a nice to have rather than a need to have.

**Comment by jam7** [comment]
I'll check it later

2017年5月31日(水) 10:45 Ed Reel <notifications@github.com>:

**Comment by jam7** [comment]
I could compile ffmpeg without libvpx well.

I experienced another error, libx264.rb checksum mismatch, though.  Can you check it your side please?

**Comment by uberhacker** [comment]
@jam7: You were right.  The checksum was incorrect.  I have pushed up a commit to fix.  I also removed the libvpx option and package as well.  If these changes compile cleanly for you, I see no reason why this cannot be merged into master.

**Pull request merged**